File: stone_lib/src/main/java/com/github/klee0kai/stone/_hidden_/IModule.java
package com.github.klee0kai.stone._hidden_;

import com.github.klee0kai.stone._hidden_.types.SwitchCacheParams;

import java.util.Set;

/**
 * Stone Private class
 */
public interface IModule {

    /**
     * Init module
     *
     * @param ob can be:
     *           - a factory instance
     *           - a factory class
     */
    boolean __init(Object ob);

    /**
     * Init caches from module prototype.
     * using in extOf method
     */
    void __initCachesFrom(IModule module);

    /**
     * Update values of bindInstance variables
     *
     * @param module related module, source to update
     */
    void __updateBindInstancesFrom(IModule module);

    /**
     * bind instance objects
     *
     * @param object - An instance of bindable objects
     */
    boolean __bind(Object object);

    /**
     * get component's factory
     *
     * @return
     */
    Object __getFactory();

    /**
     * Switch cache type for scope
     */
    void __switchRef(Set<Class> scopes, SwitchCacheParams param);

    /**
     * Clear null refs.
     * Useful after gc
     */
    void __clearNulls();


}

File: stone_multiplatform/src/commonMain/kotlin/stone/__hidden__/IModule.kt
package stone.__hidden__

/**
 * Stone Private class
 */
interface IModule {
    /**
     * Init module
     *
     * @param ob can be:
     * - a factory instance
     * - a factory class
     */
    fun __init(ob: Any?): Boolean

    /**
     * Init caches from module prototype.
     * using in extOf method
     */
    fun __initCachesFrom(module: IModule?)

    /**
     * Update values of bindInstance variables
     *
     * @param module related module, source to update
     */
    fun __updateBindInstancesFrom(module: IModule?)

    /**
     * bind instance objects
     *
     * @param object - An instance of bindable objects
     */
    fun __bind(`object`: Any?): Boolean

    /**
     * get component's factory
     *
     * @return
     */
    fun __getFactory(): Any?

    /**
     * TODO Switch cache type for scope
     */
//    fun __switchRef(scopes: MutableSet<Class<*>?>?, param: SwitchCacheParam?)

    /**
     * Clear null refs.
     * Useful after gc
     */
    fun __clearNulls()
}
